designs/Path_Resolver_DESIGN.md

```diff
@@ -0,0 +1,118 @@
+## Path Resolver for Builders
+
+### Scope
+
+When building for a particular language, there are a few things to consider that a workflow currently already does. They are the following
+
+* Language
+* Dependency Manager
+* Application framework
+
+But what is missing is the ``path to the language executable``, that will be used for building artifacts through a workflow. 
+
+Currently this is solved by just picking up the earliest instance of ``language`` in one's PATH. This works fine if there is just one instance of the language. This does not work in the case where there are more than one language version, with different executables living alongside one another in the path.
+
+This path resolver given a language and a runtime, will attempt to resolve paths in an opioniated manner before falling back to the defaulting to the first executable in `$PATH`. The resolved path is then validated to check if the resolved path actually matches the runtime attributes specified. eg: version.
+
+### Challenges
+
+How do we come up with a mechanism that looks for different versions of the same executable in the PATH. One of the ways to look at this is to explicitly look for the runtime in the path ahead of the language.
+
+eg:
+
+```bash
+Language | Runtime
+python | python3.6, python3.7
+```
+
+In this case if our runtime is python3.6, we explicitly look for python3.6 in the path first. This also gives the flexibility to the user to easily change minor versions of the same executable, by just changing the a symlink.
+
+We will need a resolver in each workflow, on being given a runtime and language combo return a path to the executable to be used for building artifacts.
+
+### Interface
+
+We will have a path resolver class, whose functionality is inherently simple to begin with. It has a candidate list of locations to look for the executable depending upon the runtime and language, and return the path.
+
+```python
+class PathResolver(object):
+     def __init__(self, language, runtime):
+        self.language = language
+        self.runtime = runtime
+        self.executables = [self.runtime, self.language]
+        
+     def path(self):
```

Comment:
Is the path allowed to be symlinks or should it be a fully resolved path?

Reply:
symlinks should be allowed in my opinion. (helps users have a quick fix to their path, with just changing symlinks)

Reply:
Path need not be fully resolved here, the chosen path will be exec'd by the validator to make a decision on the validity of the path, and if it matches the runtime.